### ISSUE TITLE
mismatching_type_param_order false positive when preceded by nontrivial generic argument

### ISSUE THREAD

**Issue by dtolnay** [rust-lang/rust-clippy#8962]
### Lint Name

mismatching_type_param_order

### Reproducer

```rust
pub trait Trait {}

pub struct Collection<T, R>(T, R);

impl<K, V, R> Trait for Collection<(K, V), R> {}
```

```console
error: `Collection` has a similarly named generic type parameter `R` in its declaration, but in a different order
 --> src/main.rs:5:44
  |
5 | impl<K, V, R> Trait for Collection<(K, V), R> {}
  |                                            ^
  |
  = note: `-D clippy::mismatching-type-param-order` implied by `-D clippy::pedantic`
  = help: try `T`, or a name that does not conflict with `Collection`'s generic params
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#mismatching_type_param_order
```

The `R` is in the correct position though. I think this code is correct as written.

### Version

```text
rustc 1.63.0-nightly (50b00252a 2022-06-06)
binary: rustc
commit-hash: 50b00252aeb77b10db04d65dc9e12ce758def4b5
commit-date: 2022-06-06
host: x86_64-unknown-linux-gnu
release: 1.63.0-nightly
LLVM version: 14.0.5
```

**Comment by dtolnay** [comment]
Mentioning @arieluy @dswij since this lint was recently added in #8831.
